### DISCUSSION TITLE
Support s390x

### DISCUSSION THREAD

**Discussion by v1gnesh** [nrwl/nx#23573] [Feature Requests]
Hello,

Can you please guide me on what's required to get nx working in s390x, and to subsequently include the building of artifacts for this platform. @pleia2 can help arrange a GHA runner in s390x, to make it a part of your regular CI.

Background: Trying to build Grafana 11 for linux/s390x - https://github.com/grafana/grafana/issues/24273#issuecomment-2114073242

**Comment by FrozenPandaz** [comment]
If I understand this correctly, you want to build a binary of Grafana which can run on s390x.
So you want to run build through Nx on a s390x machine to build the binary of Grafana?

We are working towards compiling a wasm build of the Nx binary which can be used by any system including s390x that we don't explicitly build for. #22870

We expect to have it done soon.

**Reply by v1gnesh** [reply]
>  So you want to run build through Nx on a s390x machine to build the binary of Grafana?

Yup. Grafana 11 seems to be using nx now.

I haven't dug into the CI setup of nx yet. Any pointers you can share on how I can produce something like this - 
https://www.npmjs.com/package/@nx/nx-linux-x64-gnu

That is, @nx/nx-linux-s390x-gnu ... will really help :pray:

**Comment by nealef** [comment]
I am interested in this support and would like to contribute. However, it appears that to be able to build nx you need to use nx. I have a patch-set that could be used as a starting point and builds cleanly on x64 but there are some rust compiles in there implying there is platform-specific code. This would seem to mean a cross-build would be required - unless it's just used during the build process.

I am able to set up a git actions runner for s390x if that would help. 

Going through the build steps in your instructions I end up with a local registry containing:

`README.md  node_modules  nx.json  package-lock.json  package.json  packages`

In `node_modules/nx/bin`:

`init-local.d.ts  init-local.js	nx-cloud.d.ts  nx-cloud.js  nx.d.ts  nx.js  post-install.d.ts  post-install.js	run-executor.d.ts  run-executor.js`

I have created a draft PR just so I can make my changes visible: https://github.com/nrwl/nx/pull/26772

Neale

**Reply by v1gnesh** [reply]
They're looking to do this through WASM - https://github.com/nrwl/nx/pull/22870.
There has been some related merges recently, which is good news - 
https://github.com/napi-rs/napi-rs/pull/1901
https://github.com/napi-rs/napi-rs/pull/2028

**Reply by nealef** [reply]
I cloned the repo from #22870 and checked out the `wasm` branch. I attempted to build but it is expecting to find...
```
│ /nx/node_modules/.pnpm/nx@19.2.0-rc.0_@swc-node+register@1.9.1_@swc+core@1.5.7/node_modules/nx/src/native/native-bindings.js:258
│     throw loadError
│     ^
│ Error: Cannot find module '@nx/nx-linux-s390x-gnu'
```
I take it this is the shared library created from the rust source files. However, it's a catch-22 as I can't build that unless it's already built. If there was a recipe for build this object manually I could get it in place and try the build.

**Reply by nealef** [reply]
I ran `cargo` in the `nx/src/native` directory that created `libnx.so` and `pnpm i` got further before crapping out at:
```
node_modules/.pnpm/@swc+core@1.6.6/node_modules/@swc/core: Running postinstall script, done in 51ms

> nx@0.0.1 postinstall /nx/packages/nx
> node ./bin/post-install

node:internal/modules/cjs/loader:1146
  throw err;
  ^

Error: Cannot find module '/nx/packages/nx/bin/post-install'
    at Module._resolveFilename (node:internal/modules/cjs/loader:1143:15)
    at Module._load (node:internal/modules/cjs/loader:984:27)
    at Function.executeUserEntryPoint [as runMain] (node:internal/modules/run_main:135:12)
    at node:internal/main/run_main_module:28:49 {
  code: 'MODULE_NOT_FOUND',
  requireStack: []
}
```

**Reply by nealef** [reply]
There's a `bin/post-install.ts` file so I assume I should be running `tsc` against it to create `bin/post-install.js` but I am not sure of the correct incantation nor why it's not automatically done as part of the `pnpm i` process.

**Reply by v1gnesh** [reply]
This untangling process is really challenging; multiple knots to "reverse engineer" - their CI process to see what the commands are, what the env is, how to use WASM, _its_ pre-req,  etc.

A lot of this work will become obsolete if only IBM could put some force into getting s390x CI runners supported & made available via GitHub itself... @jmertic @pleia2.
I understand you (Neil) play a pivotal part in the technology enablement of this too :)

Could you clarify please, so you've tried to a) cross-compile for s390x, from x64 b) make the wasm branch work.
Have you tried with rustc directly in linux/s390x? I find these front-end projects pretty overwhelming..

**Reply by chegde20121** [reply]
@v1gnesh I also came across same issue. I am trying to port an internal repo which uses nx on s390x. I got the below error 
`9772 verbose Linux 5.14.0-427.20.1.el9_4.s390x
9773 verbose node v16.20.2
9774 verbose npm v8.19.4
9775 error code EBADPLATFORM
9776 error notsup Unsupported platform for @nrwl/nx-linux-x64-gnu@15.8.7: wanted {"os":"linux","arch":"x64"} (current: {"os":"linux","arch":"s390x"})`

is there any solution you are aware of?

**Reply by v1gnesh** [reply]
Not yet. nx the project wants us to add support for linux/s390x by building via their wasm route.

**Reply by jmertic** [reply]
Hi all! While we don't have the resource available right now, we should later this year have an s390x Linux VM that this project can use.  More details and the request process is outlined at https://tac.openmainframeproject.org/programs/infrastructure
